File: src/services/llmService.js
/**
 * LLM Service - LiteLLM Backend Integration
 * Generates AI-powered summaries using Azure OpenAI via Webex LLM Proxy
 */

const BACKEND_URL = import.meta.env.VITE_BACKEND_URL || 'http://localhost:5001';

const requestBackendSummary = async (summaryData) => {
  const response = await fetch(`${BACKEND_URL}/api/generate-summary`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(summaryData)
  });

  if (!response.ok) {
    const errorData = await response.json().catch(() => null);
    throw new Error(errorData?.message || 'Backend API request failed');
  }

  const data = await response.json();

  if (!data.success || !data.summary) {
    throw new Error('Invalid response from backend');
  }

  return data.summary;
};

/**
 * Generate AI summary from user data
 * @param {Object} summaryData - Aggregated summary data
 * @returns {Promise<{text: string, source: string}>} AI-generated summary text
 */
export const generateSummary = async (summaryData) => {
  try {
    // Transform summaryData to match backend API format
    const backendPayload = {
      notes: {
        created: summaryData.notes.created,
        updated: summaryData.notes.updated,
        deleted: summaryData.notes.deleted
      },
      todos: {
        completed: summaryData.todos.completed,
        pending: summaryData.todos.pending
      },
      reminders: {
        triggered: summaryData.reminders.triggered,
        due: summaryData.reminders.due,
        overdue: summaryData.reminders.overdue
      }
    };

    const text = await requestBackendSummary(backendPayload);
    return {
      text,
      source: 'Azure OpenAI (LiteLLM)',
    };
  } catch (error) {
    console.warn('Backend AI summary unavailable, using fallback:', error.message);
    return {
      text: generateFallbackSummary(summaryData),
      source: 'Local Fallback',
    };
  }
};

// Prompt creation is now handled by the backend

/**
 * Generate fallback summary when API fails
 * @param {Object} data - Summary data
 * @returns {string} Template-based summary
 */
const generateFallbackSummary = (data) => {
  const stats = {
    notesCreated: data.notes.created.length,
    notesUpdated: data.notes.updated.length,
    todosCompleted: data.todos.completed.length,
    remindersDue: data.reminders.due.length,
    remindersOverdue: data.reminders.overdue.length,
    productivityScore: data.insights.productivityScore,
    streakDays: data.insights.streakDays
  };

  let summary = `📊 Activity Summary for the Last 7 Days\n\n`;

  // Overview
  summary += `You've been productive! `;
  if (stats.notesCreated > 0) {
    summary += `You created ${stats.notesCreated} new ${stats.notesCreated === 1 ? 'note' : 'notes'}`;
  }
  if (stats.notesUpdated > 0) {
    summary += ` and updated ${stats.notesUpdated} existing ${stats.notesUpdated === 1 ? 'one' : 'ones'}`;
  }
  summary += `.\n\n`;

  // Todos
  if (stats.todosCompleted > 0) {
    summary += `Great job completing ${stats.todosCompleted} ${stats.todosCompleted === 1 ? 'todo' : 'todos'}! `;
  }

  // Reminders
  if (stats.remindersDue > 0) {
    summary += `You have ${stats.remindersDue} upcoming ${stats.remindersDue === 1 ? 'reminder' : 'reminders'}. `;
  }

  if (stats.remindersOverdue > 0) {
    summary += `\n\n⚠️ Don't forget about ${stats.remindersOverdue} overdue ${stats.remindersOverdue === 1 ? 'reminder' : 'reminders'}. Take a moment to review and reschedule them.`;
  } else {
    summary += `You're staying on top of your reminders!`;
  }

  // Productivity score
  summary += `\n\n`;
  if (stats.productivityScore >= 75) {
    summary += `Your productivity score is ${stats.productivityScore}/100 - excellent work! `;
  } else if (stats.productivityScore >= 50) {
    summary += `Your productivity score is ${stats.productivityScore}/100 - you're doing well! `;
  } else {
    summary += `Your productivity score is ${stats.productivityScore}/100. There's room for improvement! `;
  }

  // Streak
  if (stats.streakDays > 0) {
    summary += `You've maintained a ${stats.streakDays}-day streak of creating notes. Keep it up!`;
  }

  // Recommendations
  summary += `\n\n💡 Suggestions:\n`;
  if (stats.remindersOverdue > 0) {
    summary += `• Review and reschedule overdue reminders\n`;
  }
  if (stats.notesCreated < 3) {
    summary += `• Try to create more notes to capture your thoughts\n`;
  }
  if (stats.todosCompleted === 0 && data.todos.total > 0) {
    summary += `• Focus on completing at least one todo today\n`;
  }

  summary += `\nKeep up the great work! 🎉`;

  return summary;
};

/**
 * Test API connection
 * @returns {Promise<boolean>} True if API is working
 */
export const testAPIConnection = async () => {
  try {
    const response = await fetch(`${BACKEND_URL}/api/test-connection`);
    const data = await response.json();
    return data.success === true;
  } catch (error) {
    console.warn('Backend API test failed:', error.message);
    return false;
  }
};

/**
 * Get API key status
 * @returns {Object} Status information
 */
export const getAPIKeyStatus = async () => {
  try {
    const response = await fetch(`${BACKEND_URL}/api/token-status`);
    const data = await response.json();
    return {
      configured: data.token_configured,
      valid: data.token_configured && data.endpoint_configured,
      endpoint: data.endpoint,
      deployment: data.deployment
    };
  } catch (error) {
    console.warn('Failed to get backend status:', error.message);
    return {
      configured: false,
      valid: false,
      endpoint: 'Unknown',
      deployment: 'Unknown'
    };
  }
};

File: backend/src/llm_service.py
"""
LLM Service using LiteLLM with Azure OpenAI via Webex
Enterprise-grade AI integration with unified interface
"""
import os
from typing import Dict, Any
import litellm
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

# Configure LiteLLM with Azure OpenAI via Webex
os.environ["AZURE_OPENAI_API_KEY"] = os.getenv("WEBEX_BEARER_TOKEN", "")
os.environ["AZURE_API_BASE"] = os.getenv("AZURE_OPENAI_ENDPOINT", "")
os.environ["AZURE_API_VERSION"] = os.getenv("AZURE_API_VERSION", "2024-10-21")

# Deployment configuration
DEPLOYMENT_NAME = os.getenv("AZURE_OPENAI_DEPLOYMENT_NAME", "gpt-4")
MODEL = f"azure/{DEPLOYMENT_NAME}"

# Enable LiteLLM debug logging (optional)
# litellm.set_verbose = True


def generate_summary(summary_data: Dict[str, Any]) -> Dict[str, Any]:
    """
    Generate AI summary using LiteLLM with Azure OpenAI backend

    Args:
        summary_data: Dictionary containing user's note activity data

    Returns:
        Dictionary with summary text and source indicator
    """
    try:
        # Create prompt from data
        prompt = create_prompt(summary_data)

        # Call LiteLLM (routes to Azure OpenAI via Webex)
        response = litellm.completion(
            model=MODEL,
            messages=[
                {
                    "role": "system",
                    "content": "You are a helpful productivity assistant analyzing user's note-taking activity. Provide friendly, actionable insights in a conversational tone."
                },
                {
                    "role": "user",
                    "content": prompt
                }
            ],
            temperature=0.7,
            max_tokens=600,
            timeout=20  # 20 second timeout
        )

        # Extract text from LiteLLM response
        summary_text = response.choices[0].message.content
        return {
            "text": summary_text,
            "is_ai": True
        }

    except litellm.exceptions.Timeout:
        print("❌ LiteLLM request timeout")
        return {
            "text": generate_fallback_summary(summary_data),
            "is_ai": False
        }

    except litellm.exceptions.AuthenticationError as e:
        print(f"❌ Authentication failed: {str(e)}")
        print("   Check your Webex bearer token in .env file")
        return {
            "text": generate_fallback_summary(summary_data),
            "is_ai": False
        }

    except litellm.exceptions.RateLimitError as e:
        print(f"❌ Rate limit exceeded: {str(e)}")
        return {
            "text": generate_fallback_summary(summary_data),
            "is_ai": False
        }

    except Exception as e:
        error_type = type(e).__name__
        print(f"❌ LiteLLM error ({error_type}): {str(e)}")
        return {
            "text": generate_fallback_summary(summary_data),
            "is_ai": False
        }


def create_prompt(data: Dict[str, Any]) -> str:
    """
    Create a detailed prompt from summary data
    """
    notes = data.get('notes', {})
    todos = data.get('todos', {})
    reminders = data.get('reminders', {})

    # Extract counts
    notes_created = len(notes.get('created', []))
    notes_updated = len(notes.get('updated', []))
    notes_deleted = len(notes.get('deleted', []))
    todos_completed = len(todos.get('completed', []))
    todos_pending = todos.get('pending', 0)
    reminders_triggered = len(reminders.get('triggered', []))
    reminders_due = len(reminders.get('due', []))
    reminders_overdue = len(reminders.get('overdue', []))

    # Get recent note titles
    recent_titles = [n.get('title', 'Untitled') for n in notes.get('created', [])[:5]]

    prompt = f"""Analyze this user's note-taking activity from the last 7 days and provide a friendly, insightful summary.

📊 Activity Summary:
- Notes created: {notes_created}
- Notes updated: {notes_updated}
- Notes deleted: {notes_deleted}
- Todos completed: {todos_completed}
- Todos pending: {todos_pending}
- Reminders triggered: {reminders_triggered}
- Upcoming reminders: {reminders_due}
- Overdue reminders: {reminders_overdue}

📝 Recent note titles:
{chr(10).join(f'- {title}' for title in recent_titles if title)}

Please provide:

1. **Overview** (2-3 sentences): Summarize their overall productivity and activity level

2. **Insights** (2-3 sentences): Identify patterns, strengths, or areas that need attention

3. **Recommendations** (2-3 sentences): Provide actionable, encouraging suggestions to improve productivity

Keep the tone:
- Conversational and friendly
- Positive and encouraging
- Practical and actionable
- Use emojis sparingly (1-2 max)

Format as plain text paragraphs separated by blank lines."""

    return prompt


def generate_fallback_summary(data: Dict[str, Any]) -> str:
    """
    Generate a simple template-based summary if LiteLLM fails
    """
    notes = data.get('notes', {})
    todos = data.get('todos', {})
    reminders = data.get('reminders', {})

    notes_created = len(notes.get('created', []))
    notes_updated = len(notes.get('updated', []))
    todos_completed = len(todos.get('completed', []))
    reminders_due = len(reminders.get('due', []))
    reminders_overdue = len(reminders.get('overdue', []))

    summary = f"""📊 Activity Summary (Last 7 Days)

You've been productive this week! You created {notes_created} new notes and updated {notes_updated} existing ones.

{"Great job completing " + str(todos_completed) + " todos!" if todos_completed > 0 else "Keep working on those todos!"}

You have {reminders_due} upcoming reminders. {f"Don't forget about {reminders_overdue} overdue reminders!" if reminders_overdue > 0 else "Great job staying on top of your reminders!"}

Keep up the good work! Stay organized and you'll continue to be productive. 🎉"""

    return summary


def test_connection() -> Dict[str, Any]:
    """
    Test LiteLLM connection to Azure OpenAI via Webex
    """
    try:
        response = litellm.completion(
            model=MODEL,
            messages=[{"role": "user", "content": "Hello, this is a test. Reply with 'Connection successful' in 3 words."}],
            max_tokens=20,
            timeout=10
        )

        return {
            "success": True,
            "message": "LiteLLM connection successful via Webex LLM Proxy",
            "model": MODEL,
            "response": response.choices[0].message.content,
            "endpoint": os.getenv("AZURE_OPENAI_ENDPOINT")
        }
    except Exception as e:
        return {
            "success": False,
            "message": f"Connection failed: {str(e)}",
            "model": MODEL,
            "endpoint": os.getenv("AZURE_OPENAI_ENDPOINT"),
            "error_type": type(e).__name__
        }


# Test on module import (for debugging)
if __name__ == "__main__":
    print("🧪 Testing LiteLLM connection to Azure OpenAI via Webex...")
    result = test_connection()
    print(f"✅ Success: {result['success']}")
    print(f"📝 Message: {result['message']}")
    if result['success']:
        print(f"💬 Response: {result['response']}")
    else:
        print(f"❌ Error Type: {result.get('error_type', 'Unknown')}")
